File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p02616/p02616_num7_parsed.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int MOD = 1e9+7;

int main() {
    int N, K;
    cin >> N >> K;
    
    vector<int> A(N);
    for (int i = 0; i < N; i++) {
        cin >> A[i];
    }
    
    sort(A.begin(), A.end());
    
    long long prod = 1;
    
    if (A[N-1] < 0 && K % 2 == 1) {
        for (int i = N-1; i >= N-K; i--) {
            prod = (prod * A[i]) % MOD;
        }
    } else {
        int left = 0, right = N-1;
        if (K % 2 == 1) {
            prod = A[right] % MOD;
            right--;
            K--;
        }
        while (K > 0) {
            long long prod_left = (long long) A[left] * A[left+1];
            long long prod_right = (long long) A[right] * A[right-1];
            
            if (prod_left > prod_right) {
                prod = (prod * prod_left) % MOD;
                left += 2;
                K -= 2;
            } else {
                prod = (prod * A[right]) % MOD;
                right--;
                K--;
            }
        }
    }
    
    cout << (prod + MOD) % MOD << endl;
    
    return 0;
}

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p02616/p02616_num7_parsed.py
import sys

MOD = 10**9 + 7

def main():
    # Read inputs
    N, K = map(int, input().split())
    A = list(map(int, input().split()))

    # Sort the list in ascending order
    A.sort()

    # If K is equal to N, then the maximum product is the product of all elements modulo MOD
    if K == N:
        product = 1
        for num in A:
            product *= num % MOD
            product %= MOD
        print(product)
        return

    # If K is odd and the first element is positive, it should be included in the product
    if K % 2 == 1 and A[-1] > 0:
        product = A[-1] % MOD
        A.pop()
        K -= 1

    # Initialize pointers
    left = 0
    right = N - 1

    # Accumulate the maximum product from both ends
    product = 1
    while K > 0:
        # Take the larger absolute value from both ends
        if abs(A[left]) >= abs(A[right]):
            product *= A[left] % MOD
            left += 1
        else:
            product *= A[right] % MOD
            right -= 1

        # Update K
        K -= 1

    print(product % MOD)

if __name__ == "__main__":
    main()